torch/csrc/jit/interpreter.cpp

```diff
@@ -19,6 +19,270 @@ namespace py = pybind11;
 
 namespace torch { namespace jit {
 
+
+// Before we translate to intepreter instructions, we do
+// some preprocessing of the graph to turn it into a form that is closer
+// to what the instructions will look like.
+// In particular we:
+// * (TODO) desugar Loop trip counts into c = 0, c += 1 instructions in the loop
+// * flatten stages so that each stage starts with a load from the stack
+//   and ends with a store to the stack
+// *. computes move_flags (see Outputs), and inserts
+//    'Drop' nodes where necessary to release Tensors after a If/Loop body exits
+//    or if a value is never used.
+// Outputs are:
+// * graph - the post processed copy of g
+// * move_flags[n] - a list of booleans, one for each input,
+//   indicating whether this is the last use of the value. The interpreter
+//   should generate a move rather than a copy in this case.
+// * stage_input_types: the type annotations on the inputs to each stage
+//   these can be removed once the the backward tracer is no longer used
+
+struct PreprocessGraph {
+  PreprocessGraph(Graph & g)
+  : graph(g.copy()) {
+    desugarTripCounts(graph->block());
+    flattenStages();
+    dropUnused(graph->block());
+
+    // fill in move_flags by scanning blocks;
+    findLastUses(graph->block());
+    //TODO: desugar Loop trip counts, for now we drop trip counts
+  }
+  std::shared_ptr<Graph> graph;
+  // for each input, should we move rather than copy the inputs
+  std::unordered_map<Node*, std::vector<uint8_t>> move_flags;
```

Comment:
We should probably canonicalize the graphs and flatten all those helper data structures to be arrays indexed by node ids